.github/workflows/ci.yml

```diff
@@ -5,21 +5,51 @@ on:
   pull_request:
 
 env:
-  CONSUL_VERSION: 1.6.10
+  DOTNET_SKIP_FIRST_TIME_EXPERIENCE: true
+  DOTNET_CLI_TELEMETRY_OPTOUT: true
 
 jobs:
 
+  # Setup our build matrix
+  setup-build-matrix:
+    if: github.event_name == 'push' || github.event.pull_request.head.repo.id != github.event.pull_request.base.repo.id
+    runs-on: ubuntu-18.04
+    outputs:
+      matrix: ${{ steps.build-matrix.outputs.matrix }}
+    steps:
+      - id: build-matrix
+        uses: jgiannuzzi/setup-build-matrix@v1
```

Comment:
@jgiannuzzi Can you recall why we need a custom action for setting up the build matrix here?

Reply:
The idea was to be able to add some combinations to the build matrix that are very difficult to express with a regular GHA matrix (they require a lot of repetition and this custom action makes it easier to manage and maintain).

Reply:
Hmm.. so does this mean we have to maintain https://github.com/jgiannuzzi/setup-build-matrix? It'd be best if we could use the default provided tooling if possible IMO. Less code to maintain the better.

Do you have an old branch with an attempt at building our matrix with the provided GitHub actions? I'd be curious to see the diff.

Reply:
Here is a before and after dating back from my exploration phase.

### regular GHA matrix

https://github.com/jgiannuzzi/consuldotnet/blob/1e9dcdf0713d3a21ac322ace0b27579dd1ffbcc6/.github/workflows/ci.yml#L10-L40
```yaml
      matrix:
        os: [ubuntu-18.04, windows-2019, macos-10.15]
        component: [Consul, Consul.AspNetCore]
        consul: [1.6.10, 1.7.10, 1.8.6, 1.9.0]
        dotnet: [2.1.811, 3.1.404, 5.0.101]
        exclude:
          - component: Consul.AspNetCore
            consul: 1.7.10
          - component: Consul.AspNetCore
            consul: 1.8.6
          - component: Consul.AspNetCore
            consul: 1.9.0
          - component: Consul.AspNetCore
            dotnet: 2.1.811
        include:
          - os: windows-2019
            component: Consul
            dotnet: "4.7"
            consul: 1.6.10
          - os: windows-2019
            component: Consul
            dotnet: "4.7"
            consul: 1.7.10
          - os: windows-2019
            component: Consul
            dotnet: "4.7"
            consul: 1.8.6
          - os: windows-2019
            component: Consul
            dotnet: "4.7"
            consul: 1.9.0
```
Job output: https://github.com/jgiannuzzi/consuldotnet/actions/runs/412890658
![image](https://user-images.githubusercontent.com/3692455/120625593-29ad6280-c462-11eb-80a1-cc3e10de152a.png)

### custom-built matrix

https://github.com/jgiannuzzi/consuldotnet/blob/2d71def285a887095004d4f7b632e7940a741b6f/.github/workflows/ci.yml#L22-L42
```yaml
            matrix:
              os: [ubuntu-18.04, windows-2019, macos-10.15]
              component: [Consul, Consul.AspNetCore]
              dotnet: [3.1.404, 5.0.101]
            operations:
              - type: merge
                match: [component]
                matrix:
                  component: [Consul]
                  consul: [1.6.10, 1.7.10, 1.8.6, 1.9.0]
              - type: add
                match: [component]
                matrix:
                  component: [Consul]
                  dotnet: [2.1.811]
              - type: add
                match: [os, component]
                matrix:
                  os: [windows-2019]
                  component: [Consul]
                  dotnet: ["4.7"]
```
Job output: https://github.com/jgiannuzzi/consuldotnet/actions/runs/415967539
![image](https://user-images.githubusercontent.com/3692455/120625412-f539a680-c461-11eb-8fdf-c190765671dd.png)

As you can see, the intent is clearer with the custom-built matrix:
- run on all OSes, for both libraries, with .NET Core 3.1 and .NET 5.0
- for the `Consul` library, test against versions 1.6 through 1.9 of the agent
- for the `Consul` library, test with .NET Core 2.1
- for the `Consul` library on Windows, test with .NET Framework 4.7

With the regular GHA matrix, multiple `exclude` and `include` tricks are used to achieve _almost_ the same result (there is a fake `consul` item with value `1.6.10` that gets created for the `Consul.AspNetCore` build). It also requires maintaining the Consul agent and .NET Framework versions in multiple places.

I delicately weighed the pros and cons of using just the default tooling and avoiding anything custom, but figured the pros outweighed the cons. Ideally this functionality would get directly built into GHA in the future, but it has to start somewhere 😄

Reply:
Thank you for the detailed explanation.

I agree that the final YML configuration is more readable and explicit with your custom action. 

However, I don't know the implications of maintaining https://github.com/jgiannuzzi/setup-build-matrix. 
For example, how often does GitHub runner updates would break our custom action? How much work will it be to fix it? What about versioning? Dependencies? Security? etc. 

[That's](https://github.com/jgiannuzzi/setup-build-matrix/blob/main/dist/index.js) 15 thousands lines of JS I'm personally unfamiliar with :-) Did you get that initial JS from somewhere? I couldn't find the default GitHub Action Matrix code anywhere (tried [this](https://github.com/marketplace?type=actions&query=matrix+), and [that](https://github.com/sdras/awesome-actions)).

Personally, I would probably stick to the default provided tooling even if the resulting config is less readable. Just my 2 cents :-)

Reply:
The actual code is in [`lib/main.js`](https://github.com/jgiannuzzi/setup-build-matrix/blob/main/lib/main.js) and is 97 lines of original hand-crafted JS 😅
They are themselves a rewrite of 46 lines of [original hand-crafted Python](https://github.com/jgiannuzzi/setup-build-matrix/blob/ecdb44368d552cf5ed422edd0d06e87d06e1b9ab/setup-build-matrix.py).

The project relies on a few JS dependencies and I should totally use dependabot to keep them up-to-date. In terms of breaking changes from GitHub Actions, the features this action relies on are probably going to be supported 'forever', as they are fundamental to most workflows. It uses the first-party language for creating actions (JS) and builds a string output from a string input. That string output is a JSON-serialised matrix, flattened to be a list KV pairs passed as `include`. GitHub would break a lot of existing workflows and actions if they changed any of that.

The GitHub Action matrix code is unfortunately not open source. [Here](https://github.com/nektos/act/blob/33ccfa6f3b85bfe02716e3369f19246de450c9e7/pkg/model/workflow.go#L196) is a compatible implementation for reference.

As the author of `setup-build-matrix`, I am of course biased, so I will let @marcin-krystianc and @mfkl decide what would be best for Consul.NET.

Reply:
Just updated the branch with Marcin's solution that uses default tooling so we can compare (previous code with your custom action [here](https://github.com/mfkl/consuldotnet/blob/mega-ci5/.github/workflows/ci.yml). 

Personally I favor using the default tooling approach.